libraries/rush-lib/src/api/RushConfiguration.ts

```diff
@@ -1753,7 +1753,10 @@ export class RushConfiguration {
    */
   public tryGetProjectForPath(currentFolderPath: string): RushConfigurationProject | undefined {
     const resolvedPath: string = path.resolve(currentFolderPath);
-    for (const project of this.projects) {
+    const sortedProjects: RushConfigurationProject[] = this.projects.sort(
```

Comment:
This is already a pretty inefficient way of searching for a project (`Path.isUnderOrEqual` can be a bit expensive). It might be a good idea to test the speed of this and make sure that it doesn't regress much, especially for the case of finding a top-level project with a short path (like apps/heft, for example).

Another way of doing this would be to create a map of packageName -> project when we instantiate the projects (minimal overhead), use `PackageJsonLookup` to locate the nearest package.json, check the map to see if the dirname is present, and if not, restart using the dirname as the `currentFolderPath`, and continue until either you find something or you're above the Rush repo directory.

You could also avoid the `Path.isUnderOrEqual` test if the `currentFolderPath` (or the dirname if it's a file, since we can't guarantee that the user passed in a folder path... though that kinda defeats much of the purpose of this shortcut) is shorter than the `project.projectFolder`. It's not ideal, but at least it avoids doing a bunch of string manip vs. just testing lengths.

Reply:
This could be more efficiently done using the LookupByPath instances that are already available through the class. Those do O(numPathSegments) lookups of the project and handle nested project folders.

Reply:
@dmichon-msft Doesn't `LookupByPath` require paths that were already normalized by Git?

Reply:
@dmichon-msft Doesn't [LookupByPath](https://github.com/microsoft/rushstack/blob/main/libraries/rush-lib/src/logic/LookupByPath.ts) assume that the path strings were already normalized by Git?

Reply:
RushConfiguration.getProjectLookupForRoot(rootPath: string) returns an optimized lookup instance that accepts paths that are relative to `rootPath`. You can invoke `lookup.findChildPathFromSegments(LookupByPath.iteratePathSegments(queryPath, path.sep))` to operate on a platform-specific relative path.

Sample code:
```ts
const resolvedPath: string = path.resolve(currentFolderPath);
// Can be any path; however, this will optimize reuse in the event that the Git root is the Rush root
const rootDirPath: string = this.rushJsonFolder;
const pathRelativeToRoot: string = path.relative(rootDirPath, resolvedPath);
// This instance will be cached for future lookups
const lookup: LookupByPath<RushConfigurationProject> = this.getProjectLookupForRoot(rootDirPath);
// Find the project by its path relative to rush.json
const result: RushConfigurationProject | undefined = lookup.findChildPathFromSegments(LookupByPath.iteratePathSegments(pathRelativeToRoot, path.sep));
return result;
```

Reply:
I have given this a test with TikTok's fairly large monorepo (437 packages) and the timing for the `tryGetProjectForPath` function call takes approximately 6-8ms, with or without the added sorting. (Tested by trying the function call on the last package listed in our rush.json file) So it does not seem like the change will cause any regressions in performance

Reply:
> RushConfiguration.getProjectLookupForRoot(rootPath: string) returns an optimized lookup instance that accepts paths that are relative to `rootPath`. You can invoke `lookup.findChildPathFromSegments(LookupByPath.iteratePathSegments(queryPath, path.sep))` to operate on a platform-specific relative path.

@dmichon-msft Even if we call `path.resolve()`, `'LIBRARIES/node-core-library/src'` will still get resolved to `'C:\\Git\\rushstack\\LIBRARIES\\node-core-library\\src'` which is going to fail the case-sensitive `Map.get()` lookup.

We can solve that, but the 6-8ms timings make it seem like not a super high priority.  My vote is to merge this PR using the simplistic `Path.isUnderOrEqual()` and then revisit the performance optimization in a separate PR.  

Right now `rush add` is not working correctly in our monorepo and we need to get people unblocked.